lib/Onyx.js

```diff
@@ -1025,17 +1028,22 @@ function initializeWithDefaultKeyStates() {
  * Storage.setItem() from Onyx.clear() will have already finished and the merged
  * value will be saved to storage after the default value.
  *
+ * @param {Array} keysToPreserve ONYXKEYS which values you want to keep
  * @returns {Promise<void>}
  */
-function clear() {
+function clear(keysToPreserve = []) {
     return getAllKeys()
         .then((keys) => {
-            _.each(keys, (key) => {
-                const resetValue = lodashGet(defaultKeyStates, key, null);
-                cache.set(key, resetValue);
-                notifySubscribersOnNextTick(key, resetValue);
+            const defaultKeys = _.keys(defaultKeyStates);
+            const keysToClear = _.without(keys, ..._.union(keysToPreserve, defaultKeys));
+            const keyValuesToClear = _.map(keysToClear, key => [key, null]);
+            const defaultKeyValues = _.pairs(_.omit(defaultKeyStates, ...keysToPreserve));
+            const newStoreKeyValues = _.union(keyValuesToClear, defaultKeyValues);
+            _.each(newStoreKeyValues, (keyValue) => {
```

Comment:
Yeah so switching this around breaks the test here:
https://github.com/Expensify/react-native-onyx/blob/335aca03d1c100f013bd3d042399e4dd25e4a066/tests/unit/onyxClearNativeStorageTest.js#L46-L66

Where the `onyxValue` ends up being the `DEFAULT_VALUE` instead. What I think is happening is that because we're calling `notifySubscribersOnNextTick` only _after_ clearing the storage that tick happens at the same time / after the merge call and this thus the `DEFAULT_VALUE` is preserved.

@neil-marcellini thoughts on this? I know you dealt with this weird race condition and `clear` before...

Reply:
Yeah so the problem we had with `Onyx.clear()` for the transition flows is that after Onyx cleared and the previous user was signed out, we would sign in the new transitioning user (with an Onyx merge), and then the default Onyx value would get set for the session key, which would sign the user out again.

So yes we need to make sure, that `clear()` followed by `merge()` on a key with a default state results in the merged value getting saved.

Reply:
OK, that's a nasty race condition, and I think `notifySubscribersOnNextTick` really bites us in that instance. It would be great to move a lot of the context from these comments into the code comment.

Reply:
@neil-marcellini feel free to take a look at the updated comment I added above this and see if you agree :+1: